cli/azd/cmd/container.go

```diff
@@ -55,20 +55,20 @@ import (
 	"golang.org/x/exp/slices"
 )
 
-// Registers a singleton action initializer for the specified action name
+// Registers a transient action initializer for the specified action name
```

Comment:
Should we move everything to transient?  I'll admit I struggle to understand the lifetime strategy, why did some of these have to change but others didn't?

Is the use of singletons buying us a to today?  Part of me thinks we should be limiting the number of singletons except in cases where they hold state that is expensive to compute or should be shared across multiple components.  That feels like very little of `azd` today (perhaps just like our tool cache abstraction?)

Reply:
Since we are introducing a workflow concept this allows users to author their own workflows.  This could then allow the user to run the same `azd` command multiple times. Consider something like

```bash
# Custom azd workflow
azd provision

azd package api
azd deploy api

azd package web
azd deploy web
```

We wouldn't want the same `package` or `deploy` commands to share any transient state like args, etc so it is safer to ensure that when the actions are created they always create a new instance.

Having `singleton` lifetime on well known components does still provider us value and gives us opportunities to reuse components, implement caching when needed. This does however require some additional cognitive load for developers to understand what the lifetime to chose when developing new components.

One such example of the above is if a workflow calls `azd package` then `azd deploy` at some point we don't want to re-package during deploy and we'd want to leverage any caching that we can to use previously packaged assets.